### PR TITLE
[PYIC-2581] Match active OAuth sessions

### PR DESCRIPTION
**Pull request by lwebb-iig**
Update `validate-cri-callback` to check if the passed OAuth `state` matches an active CRI OAuth session

### What changed

Lookup active cri oauth sessions if no ipv session provided
Throw error page if neither oauth state or ipv session not provided
Allow ipv sessions to be looked up by cri oauth session id

### Why did it change

Required to allow requests to be validated without ipv sessions

### Issue tracking

- [PYIC-2581](https://govukverify.atlassian.net/browse/PYIC-2581)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

### Other considerations

- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-2581]: https://govukverify.atlassian.net/browse/PYIC-2581?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java [hunk 1]
```diff
@@ -25,12 +25,10 @@
...
 import uk.gov.di.ipv.core.library.helpers.LogHelper;
 import uk.gov.di.ipv.core.library.helpers.StepFunctionHelpers;
+import uk.gov.di.ipv.core.library.persistence.item.ClientOAuthSessionItem;
 import uk.gov.di.ipv.core.library.persistence.item.CriOAuthSessionItem;
 import uk.gov.di.ipv.core.library.persistence.item.IpvSessionItem;
-import uk.gov.di.ipv.core.library.service.AuditService;
-import uk.gov.di.ipv.core.library.service.ConfigService;
-import uk.gov.di.ipv.core.library.service.CriOAuthSessionService;
-import uk.gov.di.ipv.core.library.service.IpvSessionService;
+import uk.gov.di.ipv.core.library.service.*;
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java]: It would be more appropriate to add only the required dependencies.

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java [hunk 2]
```diff
@@ -105,28 +107,49 @@ public Map<String, Object> handleRequest(CriCallbackRequest callbackRequest, Con
...
+            if (ipvSessionId != null) {
+                ipvSessionItem = ipvSessionService.getIpvSession(ipvSessionId);
+            } else if (criOAuthSessionId == null) {
                 throw new HttpResponseExceptionWithErrorBody(
-                        HttpStatus.SC_BAD_REQUEST, ErrorResponse.MISSING_IPV_SESSION_ID);
+                        HttpStatus.SC_BAD_REQUEST, ErrorResponse.MISSING_OAUTH_STATE);
+            } else {
+                ipvSessionItem =
+                        ipvSessionService
+                                .getIpvSessionByCriOAuthSessionId(criOAuthSessionId)
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java:122]: I think, We should be careful for performance. We can add to the new key for this property.

File: lambdas/validate-oauth-callback/src/test/java/uk/gov/di/ipv/core/credentialissuer/ValidateOAuthCallbackHandlerHandlerTest.java [hunk 3]
```diff
@@ -12,16 +12,13 @@
...
 import uk.gov.di.ipv.core.library.exceptions.SqsException;
 import uk.gov.di.ipv.core.library.helpers.SecureTokenHelper;
+import uk.gov.di.ipv.core.library.persistence.item.ClientOAuthSessionItem;
 import uk.gov.di.ipv.core.library.persistence.item.CriOAuthSessionItem;
 import uk.gov.di.ipv.core.library.persistence.item.IpvSessionItem;
-import uk.gov.di.ipv.core.library.service.AuditService;
-import uk.gov.di.ipv.core.library.service.ConfigService;
-import uk.gov.di.ipv.core.library.service.CriOAuthSessionService;
-import uk.gov.di.ipv.core.library.service.IpvSessionService;
+import uk.gov.di.ipv.core.library.service.*;
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/test/java/uk/gov/di/ipv/core/credentialissuer/ValidateOAuthCallbackHandlerHandlerTest.java]: It would be more appropriate to add only the required dependencies.

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java [hunk 4]
```diff
@@ -105,28 +107,49 @@ public Map<String, Object> handleRequest(CriCallbackRequest callbackRequest, Con
...
+                                .orElseThrow(
+                                        () ->
+                                                new HttpResponseExceptionWithErrorBody(
+                                                        HttpStatus.SC_UNAUTHORIZED,
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java]: We need to add one unit test for this error cover.

File: lambdas/validate-oauth-callback/src/test/java/uk/gov/di/ipv/core/credentialissuer/ValidateOAuthCallbackHandlerHandlerTest.java [hunk 5]
```diff
@@ -116,6 +118,8 @@ void shouldPersistAuthorizationCodeInCriOAuthSessionTable() throws Exception {
         when(mockIpvSessionService.getIpvSession(anyString())).thenReturn(ipvSessionItem);
         when(mockCriOAuthSessionService.getCriOauthSessionItem(any()))
                 .thenReturn(criOAuthSessionItem);
+        when(mockClientOAuthSessionDetailsService.getClientOAuthSession(any()))
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/test/java/uk/gov/di/ipv/core/credentialissuer/ValidateOAuthCallbackHandlerHandlerTest.java:121]: We can add verify with times for this method call

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java [hunk 6]
```diff
@@ -56,6 +56,12 @@ public IpvSessionItem getIpvSession(String ipvSessionId) {
         return dataStore.getItem(ipvSessionId);
     }
 
+    public Optional<IpvSessionItem> getIpvSessionByCriOAuthSessionId(String criOAuthSessionId) {
```
Comment (TalatAydinGDS) [inline lib/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java:59]: We can add key for this property after discuss with SRE team

File: lib/src/test/java/uk/gov/di/ipv/core/library/service/IpvSessionServiceTest.java [hunk 7]
```diff
@@ -59,6 +59,23 @@ void shouldReturnSessionItem() {
         assertEquals(ipvSessionItem.getCreationDateTime(), result.getCreationDateTime());
     }
 
+    @Test
+    void shouldReturnSessionItemByCriOAuthSessionId() {
```
Comment (TalatAydinGDS) [inline lib/src/test/java/uk/gov/di/ipv/core/library/service/IpvSessionServiceTest.java:63]: I think, we can add more than one test for null scenario

File: lib/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java (hunk 6 above)
Comment (TalatAydinGDS) [reply lib/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java:59]: We need to change in template.yaml for secondary index. Right?

**Review by TalatAydinGDS — COMMENTED** [review]
I added some comments. Please check them

**Review by TalatAydinGDS — COMMENTED** [review]
I could see some conflicts for this PR. Please review them.

File: lib/src/test/java/uk/gov/di/ipv/core/library/service/IpvSessionServiceTest.java (hunk 7 above)
Comment (lwebb-iig) [reply lib/src/test/java/uk/gov/di/ipv/core/library/service/IpvSessionServiceTest.java:63]: We don't have tests like this for `getIpvSessionByAuthorizationCode` or `getIpvSessionByAccessToken` which are similar methods - I think this is fine as is

File: deploy/template.yaml [hunk 8]
```diff
@@ -2210,6 +2212,12 @@ Resources:
               KeyType: "HASH"
           Projection:
             ProjectionType: "ALL"
+        - IndexName: "criOAuthSessionId"
```
Comment (TalatAydinGDS) [inline deploy/template.yaml:2220]: Did you test it in your stub and could see new index without table name changes vs etc with Cloudformation? I think, we need to sure about that. I wanted to remind that.

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java [hunk 9]
```diff
@@ -105,30 +111,49 @@ public Map<String, Object> handleRequest(CriCallbackRequest callbackRequest, Con
...
             String ipvSessionId = callbackRequest.getIpvSessionId();
-            if (ipvSessionId == null) {
+            String criOAuthSessionId = callbackRequest.getState();
+
+            if (ipvSessionId != null) {
+                ipvSessionItem = ipvSessionService.getIpvSession(ipvSessionId);
+            } else if (criOAuthSessionId == null) {
                 throw new HttpResponseExceptionWithErrorBody(
-                        HttpStatus.SC_BAD_REQUEST, ErrorResponse.MISSING_IPV_SESSION_ID);
+                        HttpStatus.SC_BAD_REQUEST, ErrorResponse.MISSING_OAUTH_STATE);
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java:118]: I think, we will discuss about error message description in different task. Because, this error message is not clear for this validation. MISSING_OAUTH_STATE(1030, "Missing OAuth state in Response") Client needs to understand error details from here.

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java [hunk 10]
```diff
@@ -105,30 +111,49 @@ public Map<String, Object> handleRequest(CriCallbackRequest callbackRequest, Con
...
+            } else {
+                ipvSessionItem =
+                        ipvSessionService
+                                .getIpvSessionByCriOAuthSessionId(criOAuthSessionId)
+                                .orElseThrow(
+                                        () ->
+                                                new HttpResponseExceptionWithErrorBody(
+                                                        HttpStatus.SC_BAD_REQUEST,
+                                                        ErrorResponse.MISSING_OAUTH_STATE));
```
Comment (TalatAydinGDS) [inline lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java]: I think so, before error message is clear or new one "Cri Oauth Session Id not found") for this case.  INVALID_OAUTH_STATE(1031, "Invalid OAuth State"),

File: lambdas/validate-oauth-callback/src/test/java/uk/gov/di/ipv/core/credentialissuer/ValidateOAuthCallbackHandlerHandlerTest.java (hunk 3 above)
Comment (TalatAydinGDS) [reply lambdas/validate-oauth-callback/src/test/java/uk/gov/di/ipv/core/credentialissuer/ValidateOAuthCallbackHandlerHandlerTest.java]: Please check this comment.

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java (hunk 10 above)
Comment (TalatAydinGDS) [reply lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java]: MISSING_OAUTH_STATE(1030, "Missing OAuth state in Response"),

File: deploy/template.yaml (hunk 8 above)
Comment (lwebb-iig) [reply deploy/template.yaml:2220]: See indexes https://eu-west-2.console.aws.amazon.com/dynamodbv2/home?region=eu-west-2#item-explorer?table=sessions-dev-laurencew

File: lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java (hunk 9 above)
Comment (lwebb-iig) [reply lambdas/validate-oauth-callback/src/main/java/uk/gov/di/ipv/core/validateoauthcallback/ValidateOAuthCallbackHandler.java:118]: `criOAuthSessionId` is derived from the callback state `callbackRequest.getState();` - if `criOAuthSessionId` is null, it means the OAuthState is missing from the callback request. Do you want a new error with `Missing OAuth state in Request` perhaps?

**Review by TalatAydinGDS — DISMISSED** [review]
LGTM

**Review by TalatAydinGDS — APPROVED** [review]
LGTM
